// @pivotal-tools/pipeline-builder/lib/pipeline-builder.js
#!/usr/bin/env node
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const jsyaml = require("js-yaml");
const fs = require("fs");
const path = require("path");
const shell = require("shelljs");
const TEMPLATES = path.resolve(__dirname, '../templates');
function isExecutable(f) {
    return f.endsWith('.sh');
}
class STE {
    constructor(resolver, templatesPath, outputPath) {
        this.resolver = resolver;
        this.templates = templatesPath;
        this.outputPath = outputPath;
    }
    var_re(templatePath) {
        if (templatePath.endsWith('.yml') || templatePath.endsWith('.yaml')) {
            return /\$\{(\w+)\}/g;
        }
        else if (templatePath.endsWith('.sh')) {
            return /\$\$\{(\w+)\}/g;
        }
        return null;
    }
    isTemplate(filePath) {
        return this.var_re(filePath) !== null;
    }
    process() {
        return __awaiter(this, void 0, void 0, function* () {
            yield fswalk(this.templates, '.', {
                doWithFile: (f, r) => __awaiter(this, void 0, void 0, function* () { return yield this.processFile(f, r); }),
                doWithDir: (f, r) => __awaiter(this, void 0, void 0, function* () { return yield this.processDir(f, r); }),
            });
        });
    }
    processFile(p, relativePath) {
        return __awaiter(this, void 0, void 0, function* () {
            console.log('f: ', p);
            let target = path.resolve(this.outputPath, relativePath);
            let basename = path.basename(target);
            if (basename.startsWith('_')) {
                basename = basename.substring(1);
                target = path.join(path.dirname(target), basename);
            }
            let var_re = this.var_re(p);
            if (var_re) {
                let rendered = yield this.renderTemplate(var_re, fs.readFileSync(p, 'UTF8'));
                fs.writeFileSync(target, rendered);
                if (isExecutable(p)) {
                    fs.chmodSync(target, '755');
                }
            }
            else {
                shell.cp(p, target);
            }
        });
    }
    renderTemplate(re, template) {
        return __awaiter(this, void 0, void 0, function* () {
            let varnames = new Set();
            template.replace(re, (mtch, ...more) => {
                varnames.add(more[0]);
                return mtch;
            });
            let map = new Map();
            let iter = varnames.values();
            let varname = iter.next().value;
            while (varname) {
                map.set(varname, (yield this.resolver(varname)) || '');
                varname = iter.next().value;
            }
            return template.replace(re, (mtch, ...more) => {
                let varname = more[0];
                return map.get(varname) || mtch;
            });
        });
    }
    processDir(p, relativePath) {
        return __awaiter(this, void 0, void 0, function* () {
            console.log('d: ', p);
            shell.mkdir('-p', path.resolve(this.outputPath, relativePath));
        });
    }
}
function fswalk(root, relative_path, handlers) {
    return __awaiter(this, void 0, void 0, function* () {
        let p = path.resolve(root, relative_path);
        const stat = fs.statSync(p);
        if (stat.isDirectory()) {
            yield handlers.doWithDir(p, relative_path);
            let entries = fs.readdirSync(p);
            for (let index = 0; index < entries.length; index++) {
                const name = entries[index];
                let relativeChild = path.join(relative_path, name);
                yield fswalk(root, relativeChild, handlers);
            }
        }
        else if (stat.isFile()) {
            yield handlers.doWithFile(p, relative_path);
        }
    });
}
function readValuesYaml(valuesFile) {
    if (fs.existsSync(valuesFile) && fs.statSync(valuesFile).isFile()) {
        return jsyaml.safeLoad(fs.readFileSync(valuesFile, 'UTF8'));
    }
    return {};
}
function storeValuesYaml(valuesFile, values) {
    fs.writeFileSync(valuesFile, jsyaml.safeDump(values));
}
function interactiveResolver(defaults, question) {
    let alreadyAsked = new Set();
    let cache = new Map();
    function ask(varname) {
        return __awaiter(this, void 0, void 0, function* () {
            let defaultValue = yield defaults(varname);
            if (defaultValue && defaultValue.indexOf('\n') >= 0) {
                return defaultValue;
            }
            let answer = (yield question(varname, defaultValue)) || defaultValue || '';
            cache.set(varname, yield answer);
            return answer;
        });
    }
    let resolve = (varname) => __awaiter(this, void 0, void 0, function* () {
        if (!alreadyAsked.has(varname)) {
            alreadyAsked.add(varname);
            return yield ask(varname);
        }
        return cache.get(varname) || '';
    });
    let cr = resolve;
    cr.cache = cache;
    return cr;
}
function appNameFromGitRepo(repoUri) {
    if (repoUri === undefined) {
        return;
    }
    let x = repoUri;
    if (x.endsWith('.git')) {
        x = x.substring(0, x.length - '.git'.length);
    }
    let slash = x.lastIndexOf('/');
    if (slash >= 0) {
        x = x.substring(slash + 1);
    }
    return x;
}
function usernameFromGitRepo(repoUri) {
    if (repoUri === undefined) {
        return;
    }
    if (repoUri.startsWith('http')) {
        throw new Error('https git repo case not yet implemented');
    }
    let colon = repoUri.indexOf(':');
    let slash = repoUri.indexOf('/', colon);
    if (colon >= 0 && slash >= 0) {
        return repoUri.substring(colon + 1, slash);
    }
    return repoUri;
}
function escapeYaml(str) {
    if (str.indexOf('\n') >= 0) {
        return multilineYamlString(str);
    }
    return str;
}
function or(r1, r2) {
    return (self) => __awaiter(this, void 0, void 0, function* () { return (yield r1(self)) || (yield r2(self)); });
}
class RecursiceResolverBuilder {
    constructor() {
        this.resolvers = new Map();
        this.self = (name) => __awaiter(this, void 0, void 0, function* () { return yield this.resolve(name); });
        this.cache = new Map();
    }
    loadDefaults(yamlValuesFile) {
        if (fs.existsSync(yamlValuesFile)) {
            let obj = readValuesYaml(yamlValuesFile);
            for (let property in obj) {
                if (obj.hasOwnProperty(property)) {
                    this.cache.set(property, escapeYaml(obj[property]));
                }
            }
        }
    }
    build() {
        return this.self;
    }
    add(name, resolveFun) {
        let existing = this.resolvers.get(name);
        if (existing) {
            resolveFun = or(resolveFun, existing);
        }
        this.resolvers.set(name, resolveFun);
    }
    resolve(name, chain) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.cache.has(name)) {
                return this.cache.get(name);
            }
            chain = chain || [];
            if (chain.includes(name)) {
                throw new Error(`Cyclic dependency detected: ${name} -> ${chain}`);
            }
            chain = chain.concat([name]);
            let propertyResolver = this.resolvers.get(name);
            let resolved = undefined;
            if (!propertyResolver) {
                resolved = `CHANGEME_${name}`;
            }
            else {
                resolved = yield propertyResolver(this.self);
            }
            console.log(`${name} => '${resolved}'`);
            if (resolved) {
                this.cache.set(name, resolved);
            }
            return resolved;
        });
    }
}
function tagForBranch(branch) {
    if (branch === 'master') {
        return 'latest';
    }
    return branch || 'master';
}
function exec(command) {
    return shell.exec(command, { silent: true }).stdout.toString();
}
function getDockerUser() {
    let dockerInfo = exec('docker info');
    let userPattern = /Username: (.*)/;
    let match = userPattern.exec(dockerInfo);
    return (match && match[1].trim()) || 'docker_user';
}
function multilineYamlString(str) {
    return '|\n  ' + str.replace(/\n/gm, '\n  ').trim();
}
function generate_pipeline(projectRoot, questioner) {
    return __awaiter(this, void 0, void 0, function* () {
        let defaults = new RecursiceResolverBuilder();
        defaults.loadDefaults('ci/secrets.yml');
        defaults.add('git_repo_uri', (resolve) => __awaiter(this, void 0, void 0, function* () { return exec('git config --get remote.origin.url').trim(); }));
        defaults.add('https_git_repo_uri', (resolve) => __awaiter(this, void 0, void 0, function* () {
            let uri = yield resolve('git_repo_uri');
            if (uri && uri.startsWith('git@')) {
                return 'https://' + uri.substring('git@'.length).replace(':', '/');
            }
            return uri;
        }));
        defaults.add('git_branch', (resolve) => __awaiter(this, void 0, void 0, function* () { return exec('git rev-parse --abbrev-ref HEAD').trim() || 'master'; }));
        defaults.add('app_name', (resolve) => __awaiter(this, void 0, void 0, function* () { return appNameFromGitRepo(yield resolve('git_repo_uri')); }));
        defaults.add('docker_tag', (resolve) => __awaiter(this, void 0, void 0, function* () { return tagForBranch(yield resolve('git_branch')); }));
        defaults.add('docker_repo', (resolve) => __awaiter(this, void 0, void 0, function* () {
            let user = yield resolve('docker_user');
            return `${yield resolve('docker_user')}/${yield resolve('app_name')}`;
        }));
        defaults.add('docker_image', (resolve) => __awaiter(this, void 0, void 0, function* () { return `${yield resolve('docker_repo')}:${yield resolve('docker_tag')}`; }));
        defaults.add('docker_user', (resolve) => __awaiter(this, void 0, void 0, function* () { return getDockerUser(); }));
        defaults.add('git_user', (resolve) => __awaiter(this, void 0, void 0, function* () { return usernameFromGitRepo(yield resolve('git_repo_uri')); }));
        defaults.add('helm_release_name', (resolve) => __awaiter(this, void 0, void 0, function* () { return (yield resolve('app_name')) + '-' + (yield resolve('git_branch')); }));
        defaults.add('pipeline_name', (resolve) => __awaiter(this, void 0, void 0, function* () { return (yield resolve('app_name')) + '-' + (yield resolve('git_branch')); }));
        defaults.add('kube_config', (resolve) => __awaiter(this, void 0, void 0, function* () {
            let home = process.env.HOME;
            if (home) {
                let kube_config_file = path.resolve(home, '.kube', 'config');
                if (fs.existsSync(kube_config_file)) {
                    return multilineYamlString(shell.cat(kube_config_file).toString());
                }
            }
            return 'INSERT_FULL_CONTENTS_OF_KUBE_CONFIG_FILE';
        }));
        let resolver = interactiveResolver(defaults.build(), questioner);
        shell.cd(projectRoot);
        let te = new STE(resolver, TEMPLATES, projectRoot);
        yield te.process();
        console.log(`=========================`);
        resolver.cache.forEach((v, k) => {
            console.log(`${k} = '${v}'`);
        });
        console.log(`=========================`);
    });
}
exports.generate_pipeline = generate_pipeline;
